syntax = "proto3";

service MetaService {
  rpc RegisterDataNode(DataNodeInfo) returns (Ack);
  rpc Heartbeat(DataNodeHeartbeat) returns (HeartbeatResponse);
  rpc GetFileLocation(FileLocationRequest) returns (FileLocationResponse);
  rpc AllocateChunkLocation(ChunkAllocationRequest) returns (ChunkLocation);
}

service DataNodeService {
  rpc StoreChunk(ChunkData) returns (Ack);
  rpc ReadChunk(ChunkRequest) returns (ChunkData);
}

message FileLocationRequest {
  string filename = 1;
}

message FileLocationResponse {
  repeated ChunkLocation chunks = 1;
  bool found = 2;
}

message ChunkAllocationRequest {
  string filename = 1;
  int32 chunk_index = 2;
  int64 chunk_size = 3;
}

message ChunkLocation {
  string chunk_id = 1;
  repeated string datanode_addresses = 2;
}

message DataNodeInfo {
  string address = 1;
  int64 available_space = 2;
  int32 port = 3;
}

message DataNodeHeartbeat {
  string address = 1;
  repeated string stored_chunk_ids = 2;
  int64 available_space = 3;
  int32 current_load = 4;
}

message HeartbeatResponse {
  bool ok = 1;
  repeated string chunks_to_delete = 2;
}

message ChunkData {
  string chunk_id = 1;
  bytes data = 2;
}

message ChunkRequest {
  string chunk_id = 1;
}

message Ack {
  bool ok = 1;
  string message = 2;
}
